### PR TITLE
UnusedMethod: Support additional exempting method annotations

### PR DESCRIPTION
**Pull request by rickie**
This solves two open issues (which can be linked if the PR is in a good state).

### CODE REVIEW

File: core/src/main/java/com/google/errorprone/bugpatterns/UnusedMethod.java [hunk 1]
```diff
@@ -95,30 +96,68 @@ public final class UnusedMethod extends BugChecker implements CompilationUnitTre
...
+          "org.junit.jupiter.api.AfterEach",
+          "org.junit.jupiter.api.BeforeEach",
+          "org.junit.jupiter.api.RepeatedTest",
+          "org.junit.jupiter.api.Test",
+          "org.junit.jupiter.params.ParameterizedTest");
 
   /** The set of types exempting a type that is extending or implementing them. */
   private static final ImmutableSet<String> EXEMPTING_SUPER_TYPES = ImmutableSet.of();
 
+  private final ImmutableSet<String> exemptMethodAnnotations;
```
Comment (rickie) [inline core/src/main/java/com/google/errorprone/bugpatterns/UnusedMethod.java]: An open question: 
- How do we want to name this variable? It is relatively similar to the setup in the `UnusedVariable` check.

File: core/src/test/java/com/google/errorprone/bugpatterns/UnusedMethodTest.java [hunk 2]
```diff
@@ -138,6 +139,21 @@ public void exemptedByName() {
...
+    helper
+        .addSourceLines("Foo.java", "package example;", "@interface Foo {}")
+        .addSourceLines(
+            "ExemptedViaArgs.java",
+            "package example;",
+            "class ExemptedViaArgs {",
+            "  @Foo",
+            "  private void bar() {}",
+            "}")
+        .setArgs(ImmutableList.of("-XepOpt:UnusedMethod:ExemptMethodAnnotations=example.Foo"))
```
Comment (Stephan202) [inline core/src/test/java/com/google/errorprone/bugpatterns/UnusedMethodTest.java]:
```suggestion
        .setArgs("-XepOpt:UnusedMethod:ExemptMethodAnnotations=example.Foo")
```

File: core/src/test/java/com/google/errorprone/bugpatterns/UnusedMethodTest.java [hunk 3]
```diff
@@ -138,6 +139,21 @@ public void exemptedByName() {
         .doTest();
   }
 
+  @Test
+  public void exemptedByPassingViaArgs() {
```
Comment (Stephan202) [inline core/src/test/java/com/google/errorprone/bugpatterns/UnusedMethodTest.java]: How about:
```suggestion
  public void exemptedByCustomAnnotation() {
```

File: core/src/test/java/com/google/errorprone/bugpatterns/UnusedMethodTest.java [hunk 4]
```diff
@@ -138,6 +139,21 @@ public void exemptedByName() {
...
+    helper
+        .addSourceLines("Foo.java", "package example;", "@interface Foo {}")
+        .addSourceLines(
+            "ExemptedViaArgs.java",
```
Comment (Stephan202) [inline core/src/test/java/com/google/errorprone/bugpatterns/UnusedMethodTest.java]: Could do a similar rename here; will push a proposal.

File: core/src/main/java/com/google/errorprone/bugpatterns/UnusedMethod.java [hunk 5]
```diff
@@ -95,30 +96,68 @@ public final class UnusedMethod extends BugChecker implements CompilationUnitTre
...
+
+  public UnusedMethod(ErrorProneFlags errorProneFlags) {
+    ImmutableSet.Builder<String> exemptMethodAnnotations = ImmutableSet.builder();
+    errorProneFlags
+        .getList("UnusedMethod:ExemptMethodAnnotations")
+        .ifPresent(exemptMethodAnnotations::addAll);
+    this.exemptMethodAnnotations = exemptMethodAnnotations.build();
```
Comment (Stephan202) [inline core/src/main/java/com/google/errorprone/bugpatterns/UnusedMethod.java]:
```suggestion
    this.exemptMethodAnnotations =
        errorProneFlags
            .getList("UnusedMethod:ExemptMethodAnnotations")
            .map(ImmutableSet::copyOf)
            .orElseGet(ImmutableSet::of);
```

File: core/src/main/java/com/google/errorprone/bugpatterns/UnusedMethod.java (hunk 1 above)
Comment (Stephan202) [reply core/src/main/java/com/google/errorprone/bugpatterns/UnusedMethod.java]: I think we should stick with the "exempting" terminology, as the annotations "influence" whether the methods are flagged.

File: core/src/main/java/com/google/errorprone/bugpatterns/UnusedMethod.java [hunk 6]
```diff
@@ -95,30 +96,68 @@ public final class UnusedMethod extends BugChecker implements CompilationUnitTre
...
 
   /** The set of types exempting a type that is extending or implementing them. */
   private static final ImmutableSet<String> EXEMPTING_SUPER_TYPES = ImmutableSet.of();
 
+  private final ImmutableSet<String> exemptMethodAnnotations;
+
+  public UnusedMethod(ErrorProneFlags errorProneFlags) {
+    ImmutableSet.Builder<String> exemptMethodAnnotations = ImmutableSet.builder();
+    errorProneFlags
+        .getList("UnusedMethod:ExemptMethodAnnotations")
```
Comment (Stephan202) [inline core/src/main/java/com/google/errorprone/bugpatterns/UnusedMethod.java]: Arguably we can omit the second use of "Method", as that's implied.

_Edit_: but maybe later we also want to suppose class-level exempting methods, so perhaps better to keep it as-is.

**Review by Stephan202 — APPROVED** [review]
Nice! Added a commit and tweaked the suggested PR title.

**Comment by rickie** [comment]
We can squash the commits and use the suggested commit message? (current PR title)
```
UnusedMethod: Support additional exempting method annotations 
```
One proposal/thing to consider, if we change the order of `additional` and `exempting` the focus is more on the part that supports passing in extra method annotations, whereas the current one focuses on the added method annotations.

**Comment by rickie** [comment]
Fixed in #3428.

**Pull request closed without merging**